Remove tfexec, allow TF_ environment vars and log them

spikecurtis opened the pull request:
Fixes #1359 

Couple of notes:

I've refactored our use of `terraform` and in the process made some changes that I believe bring us in line with what we are actually doing:

 * Dropped calls to `terraform state pull` because I don't believe we are supporting remote terraform state; instead it is stored in Coder
 * commands are directly passed the `Context` that gets cancelled when we read a `Cancel` message off the stream.

spikecurtis committed:
Remove tfexec, allow TF_ environment vars and log them

Signed-off-by: Spike Curtis <spike@coder.com>

spikecurtis committed:
fixup: commented code, long lines

Signed-off-by: Spike Curtis <spike@coder.com>

spikecurtis committed:
Merge branch 'main' of github.com:coder/coder into spike/1359_no_tfexec

kylecarbs commented on provisioner/terraform/executor.go:
```diff
@@ -0,0 +1,338 @@
+	return resources, nil
+}
+
+func (e executor) getState(ctx context.Context) (*tfjson.State, error) {
```
This should have the `get` prefix removed to stay consistent with other similar funcs.

kylecarbs commented on provisionersdk/logger.go:
```diff
@@ -0,0 +1,71 @@
+	"github.com/coder/coder/provisionersdk/proto"
+)
+
+type Logger interface {
+	Log(l *proto.Log) error
+}
```
Is this interface primarily to merge the `Provision` and `Parse` log streams? I think abstracting this makes it confusing to follow the code, and just to save ~30 lines of code.

spikecurtis replied:
Yeah, the idea is to be able to write generic logging code, like `LogWriter()` that work on either kind of stream.  This is something I found really useful when I prototyped my docker-compose provisioner, so I think it's really useful to have generic logging routines.

kylecarbs commented on provisioner/terraform/executor.go:
```diff
@@ -0,0 +1,338 @@
+	workdir    string
+}
+
+func (e executor) basicEnv() []string {
+	// Required for "terraform init" to find "git" to
+	// clone Terraform modules.
+	env := os.Environ()
+	// Only Linux reliably works with the Terraform plugin
+	// cache directory. It's unknown why this is.
+	if e.cachePath != "" && runtime.GOOS == "linux" {
+		env = append(env, "TF_PLUGIN_CACHE_DIR="+e.cachePath)
+	}
+	return env
+}
```
We seem to use this in every call. Instead of piping it through the execution functions, we could have a single `command` function that returns a command initialized with the `binaryPath`, `args`, and this env already applied.

spikecurtis replied:
`plan()` and `apply()` have their own environment that include parameters from end users and the coder terraform provider stuff, so not every call has the same environment in the code I'm refactoring.  I _could_ unify things, but that ends up having to either initialize the `executor` with the parameters, or pass the parameters instead of `env` on every call.  Right now the `executor` only asks for the environment on `plan()` and `apply()` and uses this basic env for every other call.

kylecarbs replied:
Makes sense!

kylecarbs commented on provisioner/terraform/executor.go:
```diff
@@ -0,0 +1,338 @@
+	return nil
+}
+
+func (e executor) getVersion(ctx context.Context) (*version.Version, error) {
```
`get` could be removed here too for consistency with other functions.

kylecarbs commented on provisioner/terraform/provision.go:
```diff
@@ -298,141 +136,60 @@ func (t *terraform) Provision(stream proto.DRPCProvisioner_ProvisionStream) erro
+	env := os.Environ()
+	for _, e := range env {
+		if strings.HasPrefix(e, "TF_") {
+			err := logger.Log(&proto.Log{Level: proto.LogLevel_WARN, Output: "terraform environment variable: " + e})
```
Terraform variables can be applied with [environment variables](https://www.terraform.io/cli/config/environment-variables#tf_var_name), which means logs would leak secrets. We could `SplitN` by `=` to print the keys instead.

kylecarbs commented on provisioner/terraform/provision_test.go:
```diff
@@ -22,24 +22,7 @@ import (
-	`
-	t.Log(provider)
-
+func testProvisioner(t *testing.T) (context.Context, proto.DRPCProvisionerClient) {
```
I might call this `setupProvisioner` instead. This could give a false impression that this is a test.

kylecarbs commented on provisionersdk/logger.go:
```diff
@@ -0,0 +1,71 @@
+	return w, done
+}
+
+func readAndLog(logger Logger, r io.Reader, done chan<- any, level proto.LogLevel) {
```
From my read over the code, this channel is only used for testing, but I'm not sure why it's required because we already ensure the output.

spikecurtis replied:
The channel is used in product code, e.g. in `executor.go:141`.  The purpose is to allow the code that calls logging routines to ensure that the logging is finished before moving on to other things.  Without this, you could end up with out-of-order logs, or worse, a log response after a complete response.

kylecarbs commented on provisionersdk/logger_test.go:
```diff
@@ -0,0 +1,123 @@
+	require.NoError(t, err)
+}
+
+func TestParseLogger_Log(t *testing.T) {
+	t.Parallel()
+
+	mStream := new(mocks.ParseStream)
+	defer mStream.AssertExpectations(t)
+
+	l := &proto.Log{Level: proto.LogLevel_ERROR, Output: "an error"}
+	mStream.On("Send", mock.MatchedBy(withLog(l))).Return(nil)
+
+	uut := provisionersdk.NewParseLogger(mStream)
+	err := uut.Log(l)
+	require.NoError(t, err)
+}
```
It seems this log abstraction tests that: a provisioner logger sends a provision log type, and a parse logger sends a parse log type. I feel like this is a lot of code to do that. Would this be caught by many other tests in the codebase?

spikecurtis replied:
Well, the compiler actually ensures that the types match up.  This tests that a call to Log results in a call to Send and that the fields match up.  Pretty basic, I know, but it's a baseline for the test below for `LogWriter()`.

These small, basic, unit tests are important right by the code they test.  It might be the case that other tests in the codebase actually cover this function, but it breaks encapsulation to depend on that for test coverage; changes elsewhere shouldn't cause us to leave this code exposed.

kylecarbs replied:
I don't disagree, but the decomposition of this comes at a relatively high cost of maintainability. Adding ~400 lines for logging seems like a *lot* right now (I'm including mock code and tests in that count), considering what we gain from it.

I'm not convinced this abstraction reduces complexity for an implementor and seems premature when we only have a single provisioner emitting logs or using this abstraction. It's hard for me to understand as a contributor the purpose of this abstraction when it's only used in a single place.

spikecurtis replied:
lines of code is a poor metric for complexity in the first place, and counting autogenerated mocks and test code in your count is counterproductive because it penalizes code that is thoroughly tested versus code that is not.

This abstraction _increases maintainability_ because it separates concerns and keeps logging boilerplate out of the provisioner functions which bloat those functions and makes them harder to read.

kylecarbs replied:
There is _some basis_ of lines of code to complexity in solving a problem. It's not strong, but if the difference is a significant factor it can be a warning of complexity. This code is ~45 lines of code in `main` (https://github.com/coder/coder/blob/main/provisioner/terraform/provision.go#L150-L195).

There are no tests in the Terraform provider that guarantee logs are sent, so I'd be in favor of moving the testing to there for now and abstracting this once we add providers. I agree with the separation of concerns, but adding an interface that's only been implemented a single time certainly adds cognitive overhead when navigating the codebase.

There are presently no parse logs sent from the Terraform provider, so half of the purpose for this abstraction isn't actually used (apart from tests) at all. It increases maintainability assuming there are many provisioners, but we only have one.

kylecarbs replied:
I'm happy to get others' perspectives on this too, it's very possible I'm being too pedantic. I care a lot about all of the code having a clear purpose for its structure, hence my hesitation to abstract before multiple usages pop up.

mafredri replied:
I already touched on this a bit in another comment. But I kind of agree with Kyle here, at least some of the abstractions. I'm fine with the `Logger` interface (I'm not sure how we'd do it otherwise), but I think the other logic could be moved into the terraform package. It's usually best to wait for 2+ (more the merrier) use-cases before creating an abstraction, that way it's easier to reason about what the common functionality should be.

And there's no need to be afraid to copy-paste code between packages until enough use-cases surface! As saying GOes:

> [A little copying is better than a little dependency.](https://go-proverbs.github.io)

spikecurtis committed:
rename executor methods to remove get

Signed-off-by: Spike Curtis <spike@coder.com>

spikecurtis committed:
don't log terraform environment variables we don't know are safe

Signed-off-by: Spike Curtis <spike@coder.com>

spikecurtis committed:
Merge branch 'main' of github.com:coder/coder into spike/1359_no_tfexec

spikecurtis committed:
Disable linting of fake secret

Signed-off-by: Spike Curtis <spike@coder.com>

mafredri commented on provisionersdk/logger.go:
```diff
@@ -0,0 +1,71 @@
+	})
+}
+
+func LogWriter(logger Logger, level proto.LogLevel) (io.WriteCloser, <-chan any) {
```
This function seems very specific to how we are running terraform and moves some of the granularity out of terraform, whilst creating a dependency provisionersdk. Some thoughts:

- We're assuming log messages will be newline delimited (scanner)
- We cannot control log level on a per-log line basis (may be relevant for future provisioners)
- This function creates an `io.Pipe` but never closes r or w, closing w is implicitly up to the caller (the done channel makes it a bit less clear we're supposed to close)
- In acutal uses, we're relying on `exec.Command` to close w (i.e. there's a lot of implicitness going on)

mafredri reviewed (commented):
I haven't looked much at the terraform package before, so I haven't 100% reviewed all changes, but added my comments and thoughts nonetheless.

spikecurtis committed:
drop parse support and move logger into terraform package

Signed-off-by: Spike Curtis <spike@coder.com>

spikecurtis commented on provisioner/terraform/provision.go:
```diff
@@ -298,141 +136,87 @@ func (t *terraform) Provision(stream proto.DRPCProvisioner_ProvisionStream) erro
+				parts[1] = "<value redacted>"
+			}
+			err := logger.Log(&proto.Log{
+				Level:  proto.LogLevel_WARN,
```
passing terraform environment variables could alter behavior in ways we have not predicted, so that's why I picked WARN.

spikecurtis commented on provisioner/terraform/executor.go:
```diff
@@ -0,0 +1,338 @@
+
+		err = logger.Log(&proto.Log{Level: logLevel, Output: log.Message})
+		if err != nil {
+			// Not much we can do.  We can't log because logging is itself breaking!
```
I did think about that, but then you end up with two "loggers" and that's just real confusing.

spikecurtis commented:
I've dropped parse support and moved the logging stuff into the terraform package.  One consequence of this is that the unit tests end up being in the `terraform` package instead of `terraform_test` because none of the logging code is exported.  I don't personally mind a mix of internal and external tests of a package, but something to be aware of.

spikecurtis committed:
disable testpackage linter on internal package test

Signed-off-by: Spike Curtis <spike@coder.com>

spikecurtis commented on provisioner/terraform/executor_test.go:
```diff
@@ -0,0 +1,63 @@
+// nolint:testpackage
```
as in `executor_internal_test.go` or something else?

kylecarbs replied:
Yup, exactly

spikecurtis committed:
Merge branch 'main' of github.com:coder/coder into spike/1359_no_tfexec

spikecurtis committed:
Merge branch 'main' of github.com:coder/coder into spike/1359_no_tfexec
